### DISCUSSION TITLE
Is it possible to rust data type to JsObject?

### DISCUSSION THREAD

**Discussion by bubbajoe** [boa-dev/boa#3891] [Q&A] [answered]
I saw the class example [here](https://github.com/boa-dev/boa/blob/main/examples/src/bin/classes.rs), but it's not quite what I was looking for.

I am trying to return an object from a `NativeFunction` and I wanted to know if it was possible to take that rust struct data type and turn into a JsObject/JsValue?

**Comment by jedel1043** [comment]
You can construct a `JsObject` from anything that implements `Trace`. Generally, you just need to add some derives to it and add a marker if the type has other types that don't implement `Trace` but that you're sure they don't contain any `Gc` pointers.
```rust
#[derive(Trace, Finalize, JsData)]
#[boa_gc(unsafe_empty_trace)]
struct MyType {
    field1: TypeWithoutTrace,
    field2: TypeWithoutTrace,
}
```
Of course, if you can implement `Trace` on `TypeWithoutTrace` it's much better because the derive works without the `unsafe_empty_trace` marker.

After that, just construct an object with the data you require:

```rust
let data = MyType { field1: TypeWithoutTrace::new(), field2: TypeWithoutTrace::new() };
let obj = JsObject::from_proto_and_data(prototype, data)
```

You'll have to pick a prototype for the object though. Generally, `context.intrinsics().constructors().object().prototype()` works well for almost all use cases, but if you want or don't need the `Object` prototype, you can just pass `None` to it.

To get the inner data, you can do `object.downcast_ref/mut::<MyType>()` if you want a direct reference to it, or `object.downcast::<MyType>()` if you want to convert the whole `JsObject` into `JsObject<MyType>`, which gives info to `JsObject` about its inner type, making it possible to access the inner data without having to `downcast_ref/mut`.

**Reply by bubbajoe** [reply]
Hmm, I still don't see the fields; Am I missing something..

So, I have `field1: JsString`. Once I do the above and run `obj.field1`, i get `undefined`.
Additionally, JSON.stringify(...) resolves to `{}`

**Reply by jedel1043** [reply]
Oh, if you want to expose Rust data to JS code, I think [`Class`](https://docs.rs/boa_engine/latest/boa_engine/class/trait.Class.html) is better for your use case. You can initialize the inner data in the [`Class::data_constructor`](https://docs.rs/boa_engine/latest/boa_engine/class/trait.Class.html#tymethod.data_constructor) method, then add a getter for that data on the [`Class::init`](https://docs.rs/boa_engine/latest/boa_engine/class/trait.Class.html#tymethod.init) method.

**Reply by jedel1043** [reply]
By the way, the previous method didn't work because inner `JsObject` data is not exposed to JS by default, it only stores data that can be accessed again from Rust code.

**Reply by bubbajoe** [reply]
Sorry, maybe my requirements a bit confusing, I am trying to return a struct value from rust that can be used in JS from a native_function. Everything that I have tried so far results in a seemingly empty object with no fields or functions.

Or perhaps I didn't understand you correctly, but this is my most recent trial using the Person class in the example
```rust
fn native_func(_this: &JsValue, args: &[JsValue], context: &mut Context) -> JsResult<JsValue> {
    let obj = JsObject::from_proto_and_data(
        context.intrinsics().constructors().object().prototype(),
        // using the same Person class from examples
        person::Person::data_constructor(
            &JsValue::undefined(),
            &[js_string!("John").into(), JsValue::new(30)],
            context,
        )?,
    );
    Ok(obj.into())
}
```

Then i get...
```javascript
let obj = native_func()
console.log(
  res.inheritedProperty,
  res.staticProperty,
  JSON.stringify(res, null, 2)
);
// undefined undefined {}
```

**Reply by jedel1043** [reply, marked as answer]
Ahhh, right. To construct an instance of `Person` you need to call either `Class::from_data` (if you already have the data you want to initialize) or `Class::construct`, if you want to construct from JS args. That will set the correct prototype, since you're setting it to `Object` in the snippet you provided.

```rust
fn native_func(_this: &JsValue, args: &[JsValue], context: &mut Context) -> JsResult<JsValue> {
    let from_data = Person::from_data(Person::new(), context)?;
    let new_target = context.get_global_class::<Self>().expect("must have the class registered").constructor();
    let from_args = Person::construct(&new_target, &[js_string!("John").into(), JsValue::new(30)],
            context);
    // more code
}
```

**Reply by bubbajoe** [reply]
Thanks, I got it to work. Is it possible create a property which can be updated from JS.

For example, referencing the classes code from examples
```rust
const person = new Person("John", 30);
// Hello my name is John, I'm 30 years old
person.name = "Jay"
person.sayHello()
// Hello my name is Jay, I'm 30 years old
```

**Reply by jedel1043** [reply]
If what you want is to set the inner data using a setter, you can add a setter on `Class::init` and use [`JsObject::downcast_mut::<MyType>()`](https://docs.rs/boa_engine/latest/boa_engine/object/struct.JsObject.html#method.downcast_mut) to get a mutable reference to the inner data that you can modify.

On the other hand, if what you want is just to set a normal JS property on the object holding the Rust data, you can initialize those on the [`Class::object_constructor`](https://docs.rs/boa_engine/latest/boa_engine/class/trait.Class.html#method.object_constructor), which will set them as common object properties.
